### ISSUE TITLE
Hang on macOS

### ISSUE THREAD

**Issue by scoopr** [gfx-rs/wgpu#1915]
**Description**
Hang on all examples that open a window, when running on macos

**Repro steps**
`cargo run --example hello-triangle`

**Expected vs observed behavior**
Should display triangle, hangs instead

**Extra materials**
Backtrace when I break in lldb:
```
  * frame #0: 0x00007fff20311cde libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff20344e49 libsystem_pthread.dylib`_pthread_cond_wait + 1298
    frame #2: 0x00000001005650ba hello-triangle`_$LT$parking_lot_core..thread_parker..imp..ThreadParker$u20$as$u20$parking_lot_core..thread_parker..ThreadParkerT$GT$::park::h77f2595a14c83ae2(self=0x0000000101509cc8) at unix.rs:77:21
    frame #3: 0x000000010055dfca hello-triangle`parking_lot_core::parking_lot::park::_$u7b$$u7b$closure$u7d$$u7d$::hd7df0fc4533d592d(thread_data=0x0000000101509cc8) at parking_lot.rs:611:17
    frame #4: 0x000000010055c85d hello-triangle`parking_lot_core::parking_lot::park::hc1932f38097a5f23 at parking_lot.rs:183:5
    frame #5: 0x000000010055c78d hello-triangle`parking_lot_core::parking_lot::park::hc1932f38097a5f23(key=4318252280, validate=closure-0 @ 0x00007ffeefbf8ea0, before_sleep=closure-1 @ 0x00007ffeefbf8ea8, timed_out=closure-2 @ 0x00007ffeefbf8eb0, park_token=(__0 = 0), timeout=Option<std::time::Instant> @ 0x00007ffeefbf8e68) at parking_lot.rs:576
    frame #6: 0x00000001006438db hello-triangle`parking_lot::raw_mutex::RawMutex::lock_slow::hc3acf2d11d87f0a1(self=0x0000000101634cf8, timeout=Option<std::time::Instant> @ 0x00007ffeefbf9038) at raw_mutex.rs:262:17
    frame #7: 0x0000000100367e13 hello-triangle`_$LT$parking_lot..raw_mutex..RawMutex$u20$as$u20$lock_api..mutex..RawMutex$GT$::lock::hd420eab4deb64dc8(self=0x0000000101634cf8) at raw_mutex.rs:72:13
    frame #8: 0x00000001003a5835 hello-triangle`lock_api::mutex::Mutex$LT$R$C$T$GT$::lock::h24610e7918704673(self=0x0000000101634cf8) at mutex.rs:214:9
    frame #9: 0x000000010039bd9c hello-triangle`wgpu_hal::metal::surface::_$LT$impl$u20$wgpu_hal..metal..Surface$GT$::dimensions::ha97a3608dc74fbdd(self=0x0000000101634cf0) at surface.rs:188:43
    frame #10: 0x000000010039d4b7 hello-triangle`wgpu_hal::metal::surface::_$LT$impl$u20$wgpu_hal..Surface$LT$wgpu_hal..metal..Api$GT$$u20$for$u20$wgpu_hal..metal..Surface$GT$::acquire_texture::h0abf0ab40d2aa0c7(self=0x0000000101634cf0, _timeout_ms=1000) at surface.rs:281:23
    frame #11: 0x00000001001e9959 hello-triangle`wgpu_core::present::_$LT$impl$u20$wgpu_core..hub..Global$LT$G$GT$$GT$::surface_get_current_texture::hc58b21ca209723f6(self=0x0000000101827c10, surface_id=Id<wgpu_core::instance::Surface> @ 0x00007ffeefbf9e68, texture_id_in=PhantomData<wgpu_core::id::Id<wgpu_core::resource::Texture<wgpu_hal::empty::Api>>> @ 0x00007ffeefbf9e70) at present.rs:124:51
    frame #12: 0x00000001001f3400 hello-triangle`_$LT$wgpu..backend..direct..Context$u20$as$u20$wgpu..Context$GT$::surface_get_current_texture::hbcf3fb8afd72f381(self=0x0000000101827c10, surface=0x000000010146f628) at direct.rs:919:15
    frame #13: 0x00000001001fd2a9 hello-triangle`wgpu::Surface::get_current_frame::h7c23f434789beb15(self=0x000000010146f620) at lib.rs:3118:13
    frame #14: 0x000000010000999d hello-triangle`hello_triangle::run::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h918ed13c0f46604c(event=Event<()> @ 0x00007ffeefbfa7b0, (null)=0x000000010170aaa0, control_flow=0x00000001006d9670) at main.rs:95:29
    frame #15: 0x000000010000cd35 hello-triangle`_$LT$winit..platform_impl..platform..app_state..EventLoopHandler$LT$T$GT$$u20$as$u20$winit..platform_impl..platform..app_state..EventHandler$GT$::handle_nonuser_event::_$u7b$$u7b$closure$u7d$$u7d$::ha35ae020fada7a66(this=0x000000010146e6d0, callback=RefMut<FnMut<(winit::event::Event<()>, &winit::event_loop::EventLoopWindowTarget<()>, &mut winit::event_loop::ControlFlow)>> @ 0x00007ffeefbfaa28) at app_state.rs:100:13
    frame #16: 0x000000010000d06f hello-triangle`winit::platform_impl::platform::app_state::EventLoopHandler$LT$T$GT$::with_callback::h2df2f71fee8dd1d3(self=0x000000010146e6d0, f=closure-0 @ 0x00007ffeefbfaac0) at app_state.rs:78:13
    frame #17: 0x000000010000cbaa hello-triangle`_$LT$winit..platform_impl..platform..app_state..EventLoopHandler$LT$T$GT$$u20$as$u20$winit..platform_impl..platform..app_state..EventHandler$GT$::handle_nonuser_event::hae774ffb2d630c6f(self=0x000000010146e6d0, event=<unavailable>, control_flow=0x00000001006d9670) at app_state.rs:99:9
    frame #18: 0x00000001005948b1 hello-triangle`winit::platform_impl::platform::app_state::Handler::handle_nonuser_event::h19737f6d762a7d54(self=0x00000001006d9660, wrapper=EventWrapper @ 0x00007ffeefbfb570) at app_state.rs:205:21
    frame #19: 0x0000000100596b9c hello-triangle`winit::platform_impl::platform::app_state::AppState::cleared::hbad37e5a2cc96cf4(panic_info=Weak<winit::platform_impl::platform::event_loop::PanicInfo> @ 0x00007ffeefbfb150) at app_state.rs:384:17
    frame #20: 0x00000001005d678c hello-triangle`winit::platform_impl::platform::observer::control_flow_end_handler::_$u7b$$u7b$closure$u7d$$u7d$::hcbe3e33275a7d3c3(panic_info=Weak<winit::platform_impl::platform::event_loop::PanicInfo> @ 0x00007ffeefbfc0e8) at observer.rs:180:21
    frame #21: 0x00000001005d65ed hello-triangle`winit::platform_impl::platform::observer::control_flow_handler::_$u7b$$u7b$closure$u7d$$u7d$::hb1547e5c9e423fa4 at observer.rs:142:57
    frame #22: 0x0000000100579e26 hello-triangle`std::panicking::try::do_call::h92ddfb4ca58e1441(data="\xd0ǿ\xef\xfe\U0000007f") at panicking.rs:401:40
    frame #23: 0x000000010057d37b hello-triangle`__rust_try + 43
    frame #24: 0x0000000100579d7e hello-triangle`std::panicking::try::h55a252378bac1ec3(f=closure-0 @ 0x00007ffeefbfc220) at panicking.rs:365:19
    frame #25: 0x000000010058fc75 hello-triangle`std::panic::catch_unwind::hbd4e29f9849e678f(f=closure-0 @ 0x00007ffeefbfc290) at panic.rs:434:14
    frame #26: 0x000000010059ac6e hello-triangle`winit::platform_impl::platform::event_loop::stop_app_on_panic::h25e56e3ab0d3844c(panic_info=Weak<winit::platform_impl::platform::event_loop::PanicInfo> @ 0x00007ffeefbfc3e0, f=closure-0 @ 0x00007ffeefbfc590) at event_loop.rs:216:11
    frame #27: 0x00000001005d6565 hello-triangle`winit::platform_impl::platform::observer::control_flow_handler::hfd76459cac4197b3(panic_info=0x0000000101708d60, f=closure-0 @ 0x00007ffeefbfc798) at observer.rs:142:5
    frame #28: 0x00000001005d6720 hello-triangle`winit::platform_impl::platform::observer::control_flow_end_handler::h74935831cbba2036((null)=0x000000010170a9f0, activity=32, panic_info=0x0000000101708d60) at observer.rs:175:9
    frame #29: 0x00007fff2043cd01 CoreFoundation`__CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
    frame #30: 0x00007fff2043cb95 CoreFoundation`__CFRunLoopDoObservers + 543
    frame #31: 0x00007fff2043c166 CoreFoundation`__CFRunLoopRun + 1159
    frame #32: 0x00007fff2043b61c CoreFoundation`CFRunLoopRunSpecific + 563
    frame #33: 0x00007fff28681a83 HIToolbox`RunCurrentEventLoopInMode + 292
    frame #34: 0x00007fff286817e5 HIToolbox`ReceiveNextEventCommon + 587
    frame #35: 0x00007fff28681583 HIToolbox`_BlockUntilNextEventMatchingListInModeWithFilter + 70
    frame #36: 0x00007fff22c43172 AppKit`_DPSNextEvent + 864
    frame #37: 0x00007fff22c41945 AppKit`-[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1364
    frame #38: 0x00007fff22c33c69 AppKit`-[NSApplication run] + 586
    frame #39: 0x00000001005ed2f6 hello-triangle`_$LT$$LP$$RP$$u20$as$u20$objc..message..MessageArguments$GT$::invoke::h6eaaf3632b31dd30(imp=(libobjc.A.dylib`objc_msgSend), obj=0x000000010140a570, sel=Sel @ 0x00007ffeefbfde90, (null)=<unavailable>) at mod.rs:128:17
    frame #40: 0x00000001005ecaa2 hello-triangle`objc::message::platform::send_unverified::h66a893898b6f561a(obj=0x000000010140a570, sel=Sel @ 0x00007ffeefbfdf00, args=<unavailable>) at mod.rs:27:9
    frame #41: 0x00000001000037a4 hello-triangle`winit::platform_impl::platform::event_loop::EventLoop$LT$T$GT$::run_return::h5891c012c60f5b84 [inlined] objc::message::send_message::h5b50d5385a1c0e86(obj=0x000000010140a570, sel=Sel @ 0x00007ffeefbfe3a0) at mod.rs:178:5
    frame #42: 0x000000010000378c hello-triangle`winit::platform_impl::platform::event_loop::EventLoop$LT$T$GT$::run_return::h5891c012c60f5b84(self=0x00007ffeefbfe4c8, callback=<unavailable>) at event_loop.rs:177
    frame #43: 0x0000000100003a6d hello-triangle`winit::platform_impl::platform::event_loop::EventLoop$LT$T$GT$::run::h2d70597d7313a12a(self=EventLoop<()> @ 0x00007ffeefbfe4c8, callback=<unavailable>) at event_loop.rs:144:9
    frame #44: 0x000000010000e86c hello-triangle`winit::event_loop::EventLoop$LT$T$GT$::run::hd626b9fdf6e17f6a(self=<unavailable>, event_handler=<unavailable>) at event_loop.rs:154:9
    frame #45: 0x000000010000acd4 hello-triangle`hello_triangle::run::_$u7b$$u7b$closure$u7d$$u7d$::h0261a0f71b1f7007((null)=ResumeTy @ 0x00007ffeefbfeca0) at main.rs:77:5
    frame #46: 0x000000010000e140 hello-triangle`_$LT$core..future..from_generator..GenFuture$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$::poll::h83bd86b511b47776(self=Pin<&mut core::future::from_generator::GenFuture<generator-0>> @ 0x00007ffeefbfed30, cx=0x00007ffeefbfef70) at mod.rs:80:19
    frame #47: 0x0000000100005730 hello-triangle`pollster::block_on::h6c38787d48343aa5(fut=<unavailable>) at lib.rs:132:15
    frame #48: 0x0000000100006571 hello-triangle`hello_triangle::main::hae3879c2bf69beb9 at main.rs:139:9
    frame #49: 0x000000010000707e hello-triangle`core::ops::function::FnOnce::call_once::hdc797dc192080d2e((null)=(hello-triangle`hello_triangle::main::hae3879c2bf69beb9 at main.rs:132), (null)=<unavailable>) at function.rs:227:5
    frame #50: 0x000000010000f421 hello-triangle`std::sys_common::backtrace::__rust_begin_short_backtrace::h16c4c8d305725d20(f=(hello-triangle`hello_triangle::main::hae3879c2bf69beb9 at main.rs:132)) at backtrace.rs:125:18
    frame #51: 0x000000010000fd24 hello-triangle`std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::h16fb36d45966251b at rt.rs:49:18
    frame #52: 0x0000000100617b61 hello-triangle`std::rt::lang_start_internal::h4428f22d05a79c62 [inlined] core::ops::function::impls::_$LT$impl$u20$core..ops..function..FnOnce$LT$A$GT$$u20$for$u20$$RF$F$GT$::call_once::h7adc4e20a318e389 at function.rs:259:13 [opt]
    frame #53: 0x0000000100617b5d hello-triangle`std::rt::lang_start_internal::h4428f22d05a79c62 [inlined] std::panicking::try::do_call::h687af85ae86a6e14 at panicking.rs:401 [opt]
    frame #54: 0x0000000100617b5d hello-triangle`std::rt::lang_start_internal::h4428f22d05a79c62 [inlined] std::panicking::try::hf74f2790da0186f3 at panicking.rs:365 [opt]
    frame #55: 0x0000000100617b5d hello-triangle`std::rt::lang_start_internal::h4428f22d05a79c62 [inlined] std::panic::catch_unwind::hd14a420ea313edb8 at panic.rs:434 [opt]
    frame #56: 0x0000000100617b5d hello-triangle`std::rt::lang_start_internal::h4428f22d05a79c62 at rt.rs:34 [opt]
    frame #57: 0x000000010000fcfe hello-triangle`std::rt::lang_start::h362f67312ce15fa0(main=(hello-triangle`hello_triangle::main::hae3879c2bf69beb9 at main.rs:132), argc=1, argv=0x00007ffeefbff370) at rt.rs:48:5
    frame #58: 0x00000001000065c6 hello-triangle`main + 22
    frame #59: 0x00007fff2035ff3d libdyld.dylib`start + 1
```

**Platform**
This was macos 11.5.2, on a macpro with 580X.

**Comment by scoopr** [comment]
On a glance, commit before #1905 works fine.

**Comment by scoopr** [comment]
Repro also on M1 laptop

**Comment by scoopr** [comment]
Seems to be fixed by 076031c1
Do you have any idea why?

**Comment by cwfitzgerald** [comment]
@scoopr that pr fixed a bug with the metal backend, iirc when uploading compressed textures

**Comment by scoopr** [comment]
Could it affect the zeroing of textures, like perhaps the framebuffer texture? Maybe it was overrunning something in the memory?

**Comment by kvark** [comment]
There are no textures in hello-triangle :/
Perhaps there is some random memory stomp happening somewhere? Need to run the address sanitizers and valgrind.

**Comment by scoopr** [comment]
I actually tried with `-Zsanitizer=address`, but nothing seemed to show up there.

**Comment by scoopr** [comment]
This bothered me enough that I tried find some explanation.
It seems that in the non-working commits, in surface.rs `acquire_texture` calls `self.dimensions()`, which ends up in the catch-all match clause, which tried to lock the render_layer again. Why it goes like that on some machines and not on others, beats me. 

But later it got fixed by using the the extents from somewhere else, and not calling `dimensions`.
